src/main/java/ch/vorburger/exec/AtomicExecuteResultHandler.java

```diff
@@ -0,0 +1,109 @@
+ * {@link java.util.concurrent.atomic.AtomicReference}
+ * instead of three separate volatile fields like the
+ * original {@link org.apache.commons.exec.DefaultExecuteResultHandler} did.
+ *
+ * @see <a href="https://github.com/vorburger/ch.vorburger.exec/issues/108">Issue #108</a>
+ */
+public class AtomicExecuteResultHandler implements ExecuteResultHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(AtomicExecuteResultHandler.class);
+
+    /** The interval polling the result */
+    private static final int SLEEP_TIME_MS = 50;
+
+    private static class Holder {
+        /** The exit value of the finished process */
+        private final Integer exitValue;
+
+        /** Any offending exception */
+        private final ExecuteException exception;
+
+        Holder(int exitValue) {
+            this.exitValue = exitValue;
+            this.exception = null;
+        }
+
+        Holder(ExecuteException e) {
+            this.exception = e;
+            this.exitValue = null;
+        }
+
+        @Override
+        public String toString() {
+            return "exitValue=" + exitValue + ", exception=" + exception;
+        }
+    }
+
+    private final AtomicReference<Holder> holder = new AtomicReference<>();
+
+    @Override
+    public void onProcessComplete(int exitValue) {
+        Holder witness = holder.compareAndExchange(null, new Holder(exitValue));
+        if (witness != null) {
+            LOG.error("onProcessComplete({}) will throw IllegalStateException, already set: {}", exitValue, witness);
+            throw new IllegalStateException(
+                    "Ignoring exit value " + exitValue + " because result already set: " + witness);
+        }
+    }
+
+    @Override
+    public void onProcessFailed(ExecuteException e) {
+        Holder witness = holder.compareAndExchange(null, new Holder(e));
+        if (witness != null) {
+            LOG.error("onProcessFailed({}) will throw IllegalStateException, already set: {}", e, witness);
+            throw new IllegalStateException("Ignoring Exception, because result already set: " + witness, e);
+        }
+    }
+
+    public Optional<Integer> getExitValue() {
+        Holder current = holder.get();
+        return current != null ? Optional.ofNullable(current.exitValue) : Optional.empty();
+    }
+
+    public Optional<Exception> getException() {
+        Holder current = holder.get();
+        return current != null ? Optional.ofNullable(current.exception) : Optional.empty();
+    }
+
+    public void waitFor() throws InterruptedException {
+        while (holder.get() == null) {
+            Thread.sleep(SLEEP_TIME_MS);
+        }
+    }
+
+    public void waitFor(long timeoutInMS) throws InterruptedException {
+        final long until = System.currentTimeMillis() + timeoutInMS;
+        while (holder.get() == null && System.currentTimeMillis() < until) {
+            Thread.sleep(SLEEP_TIME_MS);
+        }
```

Comment:
this is a pretty minor nit, but for elapsed time it's safer to use System.nanoTime since currentTimeMillis is not monotone, and can flop around based on NTP.

Reply:
Great catch! Thanks a lot for this feedback. This seemed easy enough to just quickly fix - I have raised #118 for it.